pkg/util/ranger/detacher.go

```diff
@@ -487,7 +487,9 @@ func (d *rangeDetacher) detachCNFCondAndBuildRangeForIndex(conditions []expressi
 // excludeToIncludeForIntPoint converts `(i` to `[i+1` and `i)` to `i-1]` if `i` is integer.
 // For example, if p is `(3`, i.e., point { value: int(3), excl: true, start: true }, it is equal to `[4`, i.e., point { value: int(4), excl: false, start: true }.
 // Similarly, if p is `8)`, i.e., point { value: int(8), excl: true, start: false}, it is equal to `7]`, i.e., point { value: int(7), excl: false, start: false }.
-// If return value is nil, it means p is unsatisfiable. For example, `(MaxInt64` is unsatisfiable.
+// If return value is nil, it means p is unsatisfiable. For example, `(MaxUint64` is unsatisfiable.
+// The boundary value will be treated as the bigger type: For example, `(MaxInt64` of type KindInt64 will become `[MaxInt64+1` of type KindUint64,
+// and vice versa for `0)` of type KindUint64 will become `-1]` of type KindInt64.
 func excludeToIncludeForIntPoint(p *point) *point {
```

Comment:
Seems that this function is only used to check whether the range is a point or not.
Why it will affect table dual or not?

Reply:
> Seems that this function is only used to check whether the range is a point or not. Why it will affect table dual or not?

This function will convert the point to include mode. For example, input `c>9223372036854775807` will been changed to `c>=9223372036854775808`.
If an nil is returned, it means the condition is unsatisfiable. 

Hence the condition `c>9223372036854775807` with `c` (unsigned bigint) will return nil because of the bug.

This func is invoked in `allSinglePoints`. If nil is returned, it will simply skip this range(ignore unsatisfiable condition), the pos does not increase, which returns an empty slice.

https://github.com/pingcap/tidb/blob/d2eb9026a476dd76a3fbb9449834867c4ad2a87b/pkg/util/ranger/detacher.go#L533-L539

Function `allSinglePoints` has only been called inside `ExtractEqAndInCondition`, here an empty silce returned means the range is invalid and finally cause the TableDual Outside.

https://github.com/pingcap/tidb/blob/d2eb9026a476dd76a3fbb9449834867c4ad2a87b/pkg/util/ranger/detacher.go#L661-L670

Reply:
If there exists an interval whose length is large than 0, return nil.
I think it's ok to return nil here according to the comment.

Reply:
So this change will make the func `excludeToIncludeForIntPoint` impossible to return `nil`. I think the the codes in `allSinglePoints` should also be changed.

And the actual issue might be that: The range point may be Signed Int64 while the column is Unsigned Int64 or vice versa?

Reply:
> So actual issue might be that: The range point may be Signed Int64 while the column is Unsigned Int64 or vice versa?

Yes.